Option with string type arg does not allow string to start with '-'

According to your document, an 's' type argument to an option can start with a '-' or '--'.
This same is stated by the GetOption:Long perl parser.  And perl does support this.

The ruby-getoptions checks that argument and if it starts with '-', it treats it the argument to the option as more options, and then errors that the option-with-string option did not have an argument (if it was required).
Essentially, when  an option has a (required at least) string, then the arg after that option should not be further processed in any way, but just passed as the value for that options arg.

Given the following program:

```
#!/usr/bin/env ruby
require 'ruby-getoptions'
VERSION = "0.0.1"
options, remaining = GetOptions.parse(ARGV, {
                                        "o=s" => :o,
                                        "e=s" => :e,
                                        "b"   => :b,
                                        "passthruopts=s" => :passthruopts
                                      }, {fail_on_unknown: true})
puts options.inspect
puts remaining.inspect
```

Calling this as:
 pgrm -passthruopts "-X" -o Log foo
Results in:
[ERROR] missing argument for option 'passthruopts'!

It should be no different than this invocation with a space in front of '-X'
 pgrm -passthruopts " -X" -o Log foo
Results in:
{:passthruopts=>" -X", :o=>"Log"}
["foo"]

Note that this will also change the following:
NOW:
unix% pgrm -o O -e E foo
{:o=>"O", :e=>"E"}
["foo"]
unix% pgrm -o -e E foo
[ERROR] missing argument for option 'o'!

BUT AFTER FIX, the second case can no longer detect a missing arg to '-o', so:
unix% pgrm -o -e E foo
{:o=>"-e"}
["E", "foo"]

That is what the perl parser seems to do, and is highly desired for many cases, e.g. where an option to pgrm is used to gather options to pass to another program.
    pgrm -grepoptions "-n"  fileToProcesWithGrep

--- a/README.adoc
+++ b/README.adoc
@@ -118,8 +118,10 @@ Fail on unknown:: The opposite of the above option.
 Useful if you want to ensure there are no input mistakes and force the application to stop.
 
 Require Order:: Another method to use subcommands.
-It will place all remaining arguments in the remaining array as soon as it finds the first non option argument.
+It will stop parsing arguments as soon as it finds the first non option argument.
 That is, the first argument that doesn't start with `-` that is not an argument of a previous option.
++
+When used in conjunction with Pass Through it should also stop parsing arguments when it finds the first unknown option.
 
 Warn on unknown:: Less strict parsing of options.
 This will warn the user that the option used is not a valid option but it will not stop the rest of the program.

--- a/README.adoc
+++ b/README.adoc
@@ -117,7 +117,7 @@ Additional parsers are invoked on the remaining arguments based on the initial i
 Fail on unknown:: The opposite of the above option.
 Useful if you want to ensure there are no input mistakes and force the application to stop.
 
-Stop on command:: Another method to use subcommands.
+Require Order:: Another method to use subcommands.
 It will place all remaining arguments in the remaining array as soon as it finds the first non option argument.
 That is, the first argument that doesn't start with `-` that is not an argument of a previous option.
 
@@ -508,7 +508,7 @@ For example, you can use `v` to define `verbose` and `V` to define `Version`.
 
 `pass_through`:: disable warning on unknown option. Defalt: `false`.
 
-`stop_on_command`:: If `true`, it will place all remaining arguments in the remaining array as soon as it finds the first non option argument.
+`require_order`:: If `true`, it will place all remaining arguments in the remaining array as soon as it finds the first non option argument.
 That is, the first argument that doesn't start with `-` that is not an argument of a previous option.
 
 `mode`:: possible values are "normal", "bundling", "enforce_single_dash". See <<operation_modes>> for details.

--- a/README.adoc
+++ b/README.adoc
@@ -117,6 +117,10 @@ Additional parsers are invoked on the remaining arguments based on the initial i
 Fail on unknown:: The opposite of the above option.
 Useful if you want to ensure there are no input mistakes and force the application to stop.
 
+Stop on command:: Another method to use subcommands.
+It will place all remaining arguments in the remaining array as soon as it finds the first non option argument.
+That is, the first argument that doesn't start with `-` that is not an argument of a previous option.
+
 Warn on unknown:: Less strict parsing of options.
 This will warn the user that the option used is not a valid option but it will not stop the rest of the program.
 
@@ -503,6 +507,9 @@ For example, you can use `v` to define `verbose` and `V` to define `Version`.
 `fail_on_unknown`:: if `true`, it will abort when an unknown option is passed on the commandline. If `false` it will _WARN_ when an unknown option is passed. Default: `false`.
 
 `pass_through`:: disable warning on unknown option. Defalt: `false`.
+
+`stop_on_command`:: If `true`, it will place all remaining arguments in the remaining array as soon as it finds the first non option argument.
+That is, the first argument that doesn't start with `-` that is not an argument of a previous option.
 
 `mode`:: possible values are "normal", "bundling", "enforce_single_dash". See <<operation_modes>> for details.
 

--- a/README.adoc
+++ b/README.adoc
@@ -345,7 +345,7 @@ a| destype:: `@`
 a|*Optional with increment*, Not supported yet. Does it make sense to?
 ////
 
-Options:: `fail_on_unknown`, `pass_through`, `mode`.
+Options:: `fail_on_unknown`, `pass_through`, `mode`, `require_order`.
 
 == Description
 
@@ -512,6 +512,8 @@ For example, you can use `v` to define `verbose` and `V` to define `Version`.
 
 `require_order`:: If `true`, it will place all remaining arguments in the remaining array as soon as it finds the first non option argument.
 That is, the first argument that doesn't start with `-` that is not an argument of a previous option.
++
+When used in conjunction with `pass_through` it should also stop parsing arguments when it finds the first unknown option.
 
 `mode`:: possible values are "normal", "bundling", "enforce_single_dash". See <<operation_modes>> for details.
 

--- a/lib/ruby-getoptions.rb
+++ b/lib/ruby-getoptions.rb
@@ -229,6 +229,10 @@ private
         opt_match, @option_map = find_option_matches(options[i])
         if opt_match.nil?
           remaining_args.push orig_opt
+          if @options[:require_order]
+            remaining_args.push(*args)
+            return option_result, remaining_args, []
+          end
           return option_result, remaining_args, args
         end
         # Only pass argument to the last option in the options array

--- a/lib/ruby-getoptions.rb
+++ b/lib/ruby-getoptions.rb
@@ -203,22 +203,20 @@ private
       option_result = {}
       remaining_args = []
       while args.size > 0
-        # If we require order, then push all to remaining once we see an arg that is not an option, i.e.
-        #   does not start with '-' and is not a known option
-        isCmd, dummy = isOption?(args.flatten[0], mode)
-        ## puts "DEBUG: isCmd[0] #{isCmd[0]}  Args0: #{args[0]}"
-        if @options[:require_order] && @option_map.keys.flatten.grep(isCmd[0]).empty? && args[0][0] != '-'
-          remaining_args.push(*args)
-          return option_result, remaining_args
-        end
         arg = args.shift
         options, argument = isOption?(arg, mode)
+        @log.debug "arg: #{arg}, options: #{options}, argument: #{argument}"
         if options.size >= 1 && options[0] == '--'
           remaining_args.push(*args)
           return option_result, remaining_args
         elsif options.size >= 1
           option_result, remaining_args, args = process_option(arg, option_result, args, remaining_args, options, argument)
         else
+          # If require_order then push all to remaining once we see an arg that is not an option
+          if @options[:require_order]
+            remaining_args.push(arg, *args)
+            return option_result, remaining_args
+          end
           remaining_args.push arg
         end
       end
@@ -418,13 +416,11 @@ private
     end
 
     def self.process_desttype_arg(args, opt_match, optional, required = false)
-
-###
       # If this arg exists, is required, and is string type, just use it
-      if !args[0].nil? && @option_map[opt_match][:arg_opts][0] == 's' && !optional
+      if !args[0].nil? &&
+         @option_map[opt_match][:arg_opts][0] == 's' &&
+         !optional
         arg = process_option_type(args.shift, opt_match, optional)
-###
-
       elsif !args[0].nil? && option?(args[0])
         @log.debug "args[0] option"
         if required

--- a/lib/ruby-getoptions.rb
+++ b/lib/ruby-getoptions.rb
@@ -212,8 +212,8 @@ private
         elsif options.size >= 1
           option_result, remaining_args, args = process_option(arg, option_result, args, remaining_args, options, argument)
         else
-          # If require_order then push all to remaining once we see an arg that is not an option
-          if @options[:require_order]
+          # If stop_on_command then push all to remaining once we see an arg that is not an option
+          if @options[:stop_on_command]
             remaining_args.push(arg, *args)
             return option_result, remaining_args
           end

--- a/lib/ruby-getoptions.rb
+++ b/lib/ruby-getoptions.rb
@@ -212,8 +212,8 @@ private
         elsif options.size >= 1
           option_result, remaining_args, args = process_option(arg, option_result, args, remaining_args, options, argument)
         else
-          # If stop_on_command then push all to remaining once we see an arg that is not an option
-          if @options[:stop_on_command]
+          # If require_order then push all to remaining once we see an arg that is not an option
+          if @options[:require_order]
             remaining_args.push(arg, *args)
             return option_result, remaining_args
           end

--- a/lib/ruby-getoptions.rb
+++ b/lib/ruby-getoptions.rb
@@ -203,6 +203,14 @@ private
       option_result = {}
       remaining_args = []
       while args.size > 0
+        # If we require order, then push all to remaining once we see an arg that is not an option, i.e.
+        #   does not start with '-' and is not a known option
+        isCmd, dummy = isOption?(args.flatten[0], mode)
+        ## puts "DEBUG: isCmd[0] #{isCmd[0]}  Args0: #{args[0]}"
+        if @options[:require_order] && @option_map.keys.flatten.grep(isCmd[0]).empty? && args[0][0] != '-'
+          remaining_args.push(*args)
+          return option_result, remaining_args
+        end
         arg = args.shift
         options, argument = isOption?(arg, mode)
         if options.size >= 1 && options[0] == '--'
@@ -410,7 +418,14 @@ private
     end
 
     def self.process_desttype_arg(args, opt_match, optional, required = false)
-      if !args[0].nil? && option?(args[0])
+
+###
+      # If this arg exists, is required, and is string type, just use it
+      if !args[0].nil? && @option_map[opt_match][:arg_opts][0] == 's' && !optional
+        arg = process_option_type(args.shift, opt_match, optional)
+###
+
+      elsif !args[0].nil? && option?(args[0])
         @log.debug "args[0] option"
         if required
           return args, nil

--- a/lib/ruby-getoptions.rb
+++ b/lib/ruby-getoptions.rb
@@ -20,6 +20,8 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
+require 'logger'
+
 class GetOptions
   # argument_specification:
   # [ '',
@@ -38,14 +40,13 @@ class GetOptions
   def self.parse(args, option_map = {}, options = {})
     @options = options
     @option_map = {}
-    @level = 2
     set_logging()
-    info "input args: '#{args}'"
-    info "input option_map: '#{option_map}'"
-    info "input options: '#{options}'"
+    @log.info "input args: '#{args}'"
+    @log.info "input option_map: '#{option_map}'"
+    @log.info "input options: '#{options}'"
     @option_map = generate_extended_option_map(option_map)
     option_result, remaining_args = iterate_over_arguments(args, options[:mode])
-    debug "option_result: '#{option_result}', remaining_args: '#{remaining_args}'"
+    @log.debug "option_result: '#{option_result}', remaining_args: '#{remaining_args}'"
     @log = nil
     [option_result, remaining_args]
   end
@@ -64,6 +65,7 @@ private
     REPEAT_REGEX        = /\{(\d+)?(?:,\s?(\d+)?)?\}/
     NO_DEFINITION_REGEX = /^[=:+!]/
 
+
 # This is how the instance variable @option_map looks like:
 # @option_map:
 # {
@@ -75,24 +77,20 @@ private
 #   }
 # }
 
-    def self.info(msg)
-      STDERR.puts "INFO  |" + msg if @level <= 1
-    end
-
-    def self.debug(msg)
-      STDERR.puts "DEBUG |" + msg if @level <= 0
-    end
-
     def self.set_logging()
+      @log = Logger.new(STDERR)
+      @log.formatter = proc { |severity, datetime, progname, msg|
+        "#{severity} #{caller[3].split(':')[1]} #{msg}\n"
+      }
       case @options[:debug]
       when true
-        @level = 0
+        @log.level = Logger::DEBUG
       when 'debug'
-        @level = 0
+        @log.level = Logger::DEBUG
       when 'info'
-        @level = 1
+        @log.level = Logger::INFO
       else
-        @level = 2
+        @log.level = Logger::WARN
       end
     end
 
@@ -120,7 +118,7 @@ private
         definition_list.push(*definitions)
       end
       fail_on_duplicate_definitions(definition_list)
-      debug "opt_map: #{opt_map}"
+      @log.debug "opt_map: #{opt_map}"
       opt_map
     end
 
@@ -229,9 +227,9 @@ private
         end
         # Only pass argument to the last option in the options array
         args.unshift argument unless argument.nil? || argument == "" || i < (options.size - 1)
-        debug "new args: #{args}"
+        @log.debug "new args: #{args}"
         option_result, args = execute_option(opt_match, option_result, args)
-        debug "option_result: #{option_result}"
+        @log.debug "option_result: #{option_result}"
       end
       return option_result, remaining_args, args
     end
@@ -254,7 +252,7 @@ private
               # Update the given hash
               hash[k][:negated] = true
               local_matches.push name
-              debug "hash: #{hash}"
+              @log.debug "hash: #{hash}"
             end
           end
         end
@@ -278,14 +276,14 @@ private
         if @options[:fail_on_unknown]
           abort "[ERROR] Option '#{opt}' not found!"
         else
-          debug "Option '#{opt}' not found!"
+          @log.debug "Option '#{opt}' not found!"
           $stderr.puts "[WARNING] Option '#{opt}' not found!" unless @options[:pass_through]
           return [nil, @option_map]
         end
       elsif matches.size > 1
         abort "[ERROR] option '#{opt}' matches multiple names '#{matches.sort.inspect}'!"
       end
-      debug "matches: #{matches}"
+      @log.debug "matches: #{matches}"
       [matches[0], @option_map]
     end
 
@@ -293,13 +291,13 @@ private
     #       Fail during init and not during run time.
     def self.execute_option(opt_match, option_result, args)
       opt_def = @option_map[opt_match]
-      debug "#{opt_def[:arg_spec]}"
+      @log.debug "#{opt_def[:arg_spec]}"
       case opt_def[:arg_spec]
       when 'flag'
         if opt_def[:opt_dest].kind_of? Symbol
           option_result[opt_def[:opt_dest]] = true
         else
-          debug "Flag definition is a function"
+          @log.debug "Flag definition is a function"
           opt_def[:opt_dest].call
         end
       when 'nflag'
@@ -374,7 +372,7 @@ private
         min = opt_def[:arg_opts][2][0]
         max = opt_def[:arg_opts][2][1]
         while min > 0
-          debug "min: #{min}, max: #{max}"
+          @log.debug "min: #{min}, max: #{max}"
           min -= 1
           max -= 1
           abort "[ERROR] missing argument for option '#{opt_match[0]}'!" if args.size <= 0
@@ -387,7 +385,7 @@ private
           end
         end
         while max > 0
-          debug "min: #{min}, max: #{max}"
+          @log.debug "min: #{min}, max: #{max}"
           max -= 1
           break if args.size <= 0
           if type == Array
@@ -413,7 +411,7 @@ private
 
     def self.process_desttype_arg(args, opt_match, optional, required = false)
       if !args[0].nil? && option?(args[0])
-        debug "args[0] option"
+        @log.debug "args[0] option"
         if required
           return args, nil
         end
@@ -421,9 +419,9 @@ private
       else
         arg = process_option_type(args.shift, opt_match, optional)
       end
-      debug "arg: '#{arg}'"
+      @log.debug "arg: '#{arg}'"
       if arg.nil?
-        debug "arg is nil"
+        @log.debug "arg is nil"
         abort "[ERROR] missing argument for option '#{opt_match[0]}'!"
       end
       [args, arg]
@@ -440,11 +438,11 @@ private
       else
         abort "[ERROR] argument for option '#{opt_match[0]}' must be of type key=value!"
       end
-      debug "key: '#{key}', arg: '#{arg}'"
+      @log.debug "key: '#{key}', arg: '#{arg}'"
       arg = process_option_type(arg, opt_match, optional)
-      debug "arg: '#{arg}'"
+      @log.debug "arg: '#{arg}'"
       if arg.nil?
-        debug "arg is nil"
+        @log.debug "arg is nil"
         abort "[ERROR] missing argument for option '#{opt_match[0]}'!"
       end
       [args, arg, key]
@@ -460,7 +458,7 @@ private
 
     def self.option?(arg)
       result = !!(IS_OPTION_REGEX =~ arg)
-      debug "Is option? '#{arg}' #{result}"
+      @log.debug "Is option? '#{arg}' #{result}"
       result
     end
 

--- a/test/test-ruby-getoptions.rb
+++ b/test/test-ruby-getoptions.rb
@@ -675,7 +675,7 @@ describe GetOptions do
       ['--opt', 'arg', 'subcommand', '--arg', '--help'],
       { 'opt=s' => :opt,
         'help' => :help },
-      {:require_order => true}
+      {:stop_on_command => true}
     )
     options[:opt].must_equal 'arg'
     options[:help].must_equal nil

--- a/test/test-ruby-getoptions.rb
+++ b/test/test-ruby-getoptions.rb
@@ -675,7 +675,7 @@ describe GetOptions do
       ['--opt', 'arg', 'subcommand', '--arg', '--help'],
       { 'opt=s' => :opt,
         'help' => :help },
-      {:stop_on_command => true}
+      {:require_order => true}
     )
     options[:opt].must_equal 'arg'
     options[:help].must_equal nil

--- a/test/test-ruby-getoptions.rb
+++ b/test/test-ruby-getoptions.rb
@@ -60,6 +60,13 @@ describe GetOptions do
     )
     options[:string].must_equal 'test'
     remaining.must_equal ['Hello', 'world!']
+
+    options, remaining = GetOptions.parse(
+      ['Hello', '--string=--test', 'world!'],
+      {'string=s' => :string}
+    )
+    options[:string].must_equal '--test'
+    remaining.must_equal ['Hello', 'world!']
   end
 
   it 'should allow optional strings' do
@@ -87,7 +94,7 @@ describe GetOptions do
     lambda {
       lambda {
         GetOptions.parse(
-          ['--string', '--test'],
+          ['--string'],
           {'string=s' => :string, 'test' => :test}
         )
       }.must_raise(SystemExit)
@@ -661,6 +668,18 @@ describe GetOptions do
         ['Hello', 'world!', 'Hello']
       )
     }.must_raise(ArgumentError)
+  end
+
+  it 'should stop on subcommands' do
+    options, remaining = GetOptions.parse(
+      ['--opt', 'arg', 'subcommand', '--arg', '--help'],
+      { 'opt=s' => :opt,
+        'help' => :help },
+      {:require_order => true}
+    )
+    options[:opt].must_equal 'arg'
+    options[:help].must_equal nil
+    remaining.must_equal ['subcommand', '--arg', '--help']
   end
 
 end

--- a/test/test-ruby-getoptions.rb
+++ b/test/test-ruby-getoptions.rb
@@ -680,6 +680,16 @@ describe GetOptions do
     options[:opt].must_equal 'arg'
     options[:help].must_equal nil
     remaining.must_equal ['subcommand', '--arg', '--help']
+
+    options, remaining = GetOptions.parse(
+      ['--opt', 'arg', '--unknown', '--arg', '--help'],
+      { 'opt=s' => :opt,
+        'help' => :help },
+      {:require_order => true, :pass_through => true}
+    )
+    options[:opt].must_equal 'arg'
+    options[:help].must_equal nil
+    remaining.must_equal ['--unknown', '--arg', '--help']
   end
 
 end

